File: C++/1095-Find in Mountain Array.cpp
/**
 * // This is the MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * class MountainArray {
 *   public:
 *     int get(int index);
 *     int length();
 * };
 */

class Solution {
public:
    int findInMountainArray(int target, MountainArray &A) {
        int left = 0, right = A.length()-1;
        int peak;
        while (left + 1 < right){
            int mid = left + (right-left)/2;
            if (A.get(mid) < A.get(mid+1))
                left = mid;
            else
                right = mid;
        }
        
        peak = left;
        
        left = 0;
        right = peak;
        while (left + 1 < right){
            int mid = left + (right-left)/2;
            if (A.get(mid) <= target)
                left = mid;
            else
                right = mid;
        }        
        if (A.get(left) == target) return left;
        if (A.get(right) == target) return right;
        
        
        left = peak+1;
        right = A.length()-1;
        while (left + 1 < right){
            int mid = left + (right-left)/2;
            if (A.get(mid) >= target)
                left = mid;
            else
                right = mid;
        }        
        if (A.get(left) == target) return left;
        if (A.get(right) == target) return right;        
        
        return -1;
    }
};

File: 1095-Find in Mountain Array.py
# """
# This is MountainArray's API interface.
# You should not implement it, or speculate about its implementation
# """
#class MountainArray:
#    def get(self, index: int) -> int:
#    def length(self) -> int:

class Solution:
    def findInMountainArray(self, target: int, mountain_arr: 'MountainArray') -> int:
        N = mountain_arr.length()
        left, right = 0, N-1
        
        while left + 1 < right:
            mid = left + (right-left)//2
            if mountain_arr.get(mid) < mountain_arr.get(mid+1):
                left = mid
            else:
                right = mid
        
        peak = left
        
        l, r = 0, peak
        
        while l + 1 < r:
            mid = l + (r-l)//2
            if mountain_arr.get(mid) <= target:
                l = mid
            else:
                r = mid
        
        if mountain_arr.get(l) == target:
            return l
        if mountain_arr.get(r) == target:
            return r
        
        
        l, r = peak+1, N-1
        
        while l + 1 < r:
            mid = l + (r-l)//2
            if mountain_arr.get(mid) >= target:
                l = mid
            else:
                r = mid
        
        if mountain_arr.get(l) == target:
            return l
        if mountain_arr.get(r) == target:
            return r
                
        return -1
